addon/-private/system/store/finders.js

```diff
@@ -38,6 +38,11 @@ export function _find(adapter, store, typeClass, id, internalModel, options) {
     return store._adapterRun(function() {
       var payload = normalizeResponseHelper(serializer, store, typeClass, adapterPayload, id, 'findRecord');
       assert('Ember Data expected the primary data returned from a `findRecord` response to be an object but instead it found an array.', !Array.isArray(payload.data));
+
+      warn(`You requested a record of type '${typeClass.modelName}' with id '${id}' but the adapter returned a payload with primary data having an id of '${payload.data.id}'. Looks like you want to use store.queryRecord instead http://emberjs.com/api/data/classes/DS.Store.html#method_queryRecord`, payload.data.id === id, {
```

Comment:
Hmm, I see your concern 😔 . But I don't see how this issue could be solved other than `queryRecord`. I will try to update the docs and this warning...